### ISSUE TITLE
Getting a 412 Precondition Error on POST

### ISSUE THREAD

**Issue by bazwilliams** [PeteManchester/MediaPlayer#37]
I'm running mediaplayer on a Linux box, from a windows machine using Chrome and the 'Postman' plugin to send a SOAP request to mediaplayer I get a 412 'precondition failed'. I'm not sure which additional headers I need to satisfy the request. 

Here is my request URI:
http://192.168.1.126:52821/Ds/Product/control

Headers:
```
Accept: text/xml
Content-Type: text/xml
SOAPAction: urn:av-openhome.org:service:Product:1#SourceXml
```

Body:
```
<?xml version="1.0"?>
<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/">
  <s:Body s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/">
    <u:SourceXml xmlns:u="urn:av-openhome.org:service:Product:1">
    </u:SourceXml>
  </s:Body>
</s:Envelope>
```

**Comment by PeteManchester** [comment]
Hi Barry,

The short answer is I'm not sure, the SOAP messages are at lower level than I normally with in my MediaPlayer.

In my code I don't explicitly implement a SOAP interface and so if there is one available it is probably part of the ohNet libraries, 

It might be best to ask on the Openhome forums, I found this thread 

http://forum.openhome.org/archive/index.php?thread-1132.html

Or even using WireShark to capture the traffic to mediaplayer might give you some clues.

I would be interested to hear how you get on though, or if you need anymore help let me know.

**Comment by bazwilliams** [comment]
Cheers,

I'm using tcpdump to see what the difference between Kazoo and my app is when speaking to MediaPlayer. Only difference appears to be Kazoo making HTTP 1.0 calls and less headers as a result which is my prime suspect. 

Will try to simulate that.

**Comment by bazwilliams** [comment]
Solved!

Turns out it is nothing to do with headers, HTTP 1.0 or 1.1. I was just requesting the wrong path; if I'd read the device.xml documentation I would have seen the follow URI to use: 

http://192.168.1.126:52821/device-Ethan's%20Bedroom-zorin-MediaRenderer/av.openhome.org-Product-1/control

I guess the 412 sent me off track, I'd have expected a 404... 

I'll close this as a non issue now. I've got a node based playlist/alarm app which works with my Linn equipment, but not yet with your excellent app which I use in my son's room. 

Keep up the good work :)

**Comment by PeteManchester** [comment]
Yeah you would expect a 404, sorry I couldn't be more use..

Is that the GloryDS that is on the Linn Forum or a different one..

When you get it working for my mediaplayer can you send me a copy.

I have a simple AlarmClock plugin for the Mediaplayer but I always intended to add a web page to configure the alarms, as usual I didn't get round to doing it..

**Comment by bazwilliams** [comment]
No worries - you pointed me in the direction of sniffing the traffic which highlighted the difference!

The app I'm writing has been around for a little while and is on Github! It's Node.js based: 

https://github.com/bazwilliams/upnp-playlist-service

Originally started out as an alarm clock service for all my DSs, but slowly mutated into a playlist manager so I can save and reload playlists directly from the DS without having to use Kazoo.

**Comment by PeteManchester** [comment]
I forgot to mention, if there is anything I can change on the Mediaplayer
side to make it more compatible with your current Node.js let me know,
maybe I can change things on my side to make it more like a DS rather than
you change things on your side..

**Comment by bazwilliams** [comment]
Cool - I'll let you know if I spot anything!

**Comment by bazwilliams** [comment]
I've updated the playlist app to use the documented control Urls and I've tested the playlist load and save functionality on my MediaPlayer box. 

Let me know if you try it and feel free to suggest any improvements/fixes. 

Bit of a plug for my blog, but I use the playlist app and wireless light switch to control playlists for the Ds in my bathroom. Should now work for MediaPlayer boxes too :)

http://blog.bjw.me.uk/2015/01/linn-ds-homebrew-rcu-with-playlist.html
